src/realm/object-store/impl/deep_change_checker.cpp

```diff
@@ -23,91 +23,194 @@
+                                                std::vector<KeyPathArray>& key_path_arrays)
 {
     auto table_key = table.get_key();
-    if (any_of(begin(out), end(out), [=](auto& tbl) {
+    // If the currently looked at `table` is already part of the `std::vector<RelatedTable>` (possibly
+    // due to another path involving it) we do not need to traverse further and can return.
+    if (any_of(begin(out), end(out), [&](const auto& tbl) {
             return tbl.table_key == table_key;
         }))
         return;
 
+    // If a filter is set and the table is not part of the filter, it can be skipped.
+    if (tables_in_filters.size() != 0) {
+        if (none_of(begin(tables_in_filters), end(tables_in_filters), [&](const auto& filtered_table_key) {
+                return filtered_table_key == table_key;
+            })) {
+            return;
+        }
+    }
+
     // We need to add this table to `out` before recurring so that the check
     // above works, but we can't store a pointer to the thing being populated
     // because the recursive calls may resize `out`, so instead look it up by
-    // index every time
+    // index every time.
     size_t out_index = out.size();
     out.push_back({table_key, {}});
 
     for (auto col_key : table.get_column_keys()) {
         auto type = table.get_column_type(col_key);
+        // If a column within the `table` does link to another table it needs to be added to `table`'s
+        // links.
         if (type == type_Link || type == type_LinkList) {
             out[out_index].links.push_back({col_key.value, type == type_LinkList});
-            find_related_tables(out, *table.get_link_target(col_key));
+            // Finally this function needs to be called again to traverse all linked tables using the
+            // just found link.
+            find_all_related_tables(out, *table.get_link_target(col_key), tables_in_filters, key_path_arrays);
+        }
+    }
+    if (tables_in_filters.size() != 0) {
+        // Backlinks can only come into consideration when added via key paths.
+        // If there are not `tables_in_filter` we can skip this part.
+        table.for_each_backlink_column([&](ColKey column_key) {
+            // If this backlink is included in any of the filters we follow the path further.
+            out[out_index].links.push_back({column_key.value, false});
+            const Table& origin_table = *table.get_link_target(column_key);
+
+            for (auto& key_path_array : key_path_arrays) {
+                for (auto& key_path : key_path_array) {
+                    auto key_path_size = key_path.size();
+                    if (key_path_size > 0) {
+                        auto last_key_path_element = key_path[key_path_size - 1];
+                        auto last_table_key = last_key_path_element.first;
+                        auto last_column_key = last_key_path_element.second;
```

Comment:
This is basically the same reason as in https://github.com/realm/realm-core/pull/4646#discussion_r653847256. If there is a backlink at the end we also need the table after that. But if the backlink is in the middle of a KeyPath the origin table will be added anyway since it's the next part of the KeyPath.

I must admit, after all the refactoring I realise that might not be the right location and even be misleading or confusing. I'll see where it makes more sense in case you don't see a mistake in my thought process in general.